federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+	// which may involve service interruption.  Also, we would like user-friendly events.
+
+	// Save the state so we can avoid a write if it doesn't change
+	//TODO: jesse
+	//previousState := api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer) // TODO: quinton: are the DNS names in here?  I think so.
+	if !wantsDNSRecords(service) {
+		needDelete := true
+		if appliedState != nil {
+			if !wantsDNSRecords(appliedState) {
+				needDelete = false
+			}
+		} else {
+			// If we don't have any cached memory of the DNS records, we have to ask
+			// the cloud provider for what it knows about them.
+			// Technically EnsureDNSRecordsDeleted can cope, but we want to post meaningful events
+			//TODO: Jesse:uncomment
+			//_, exists, err := s.dns.GetDNSRecords(service)
+			//if err != nil {
+			//	return fmt.Errorf("Error getting DNS records for service %s: %v", namespacedName, err), retryable
+			//}
+			//if !exists {
+			//	needDelete = false
+			//}
+		}
+
+		if needDelete {
+			glog.Infof("Deleting existing DNS records for service %s that no longer needs them.", key)
+			s.eventRecorder.Event(service, api.EventTypeNormal, "DeletingDNSRecords", "Deleting DNS records")
+			//todo:Jesse
+			//if err := s.dns.EnsureDNSRecordsDeleted(service); err != nil {
+			//	return err, retryable
+			//}
+			s.eventRecorder.Event(service, api.EventTypeNormal, "DeletedDNSRecords", "Deleted DNS records")
+		}
+
+		service.Status.LoadBalancer = v1.LoadBalancerStatus{} // TODO: quinton: is the DNS name in here?
+	} else {
+		glog.V(2).Infof("Ensuring DNS records for service %s", key)
+
+		// TODO: We could do a dry-run here if wanted to avoid the spurious cloud-calls & events when we restart
+
+		// The DNS records don't exist yet, so create them.
+		// TODO: quinton: Split this up and handle each record separately.
+		//                Beware though, creating distant ones before near ones can lead to race condidtions on name resolution.
+		//                (e.g. client local resolution fails as DNS record doesn't exist yet, but then tries remote resolution and that succeeds,
+		//                so it gets stuck for a while on a remote service shard.
+		//                Solutions: either use transactions (where providers suport them, e.g. Google Cloud DNS) or just make
+		//                sure that the creation sequence is zone-local, region-local, sub-continent local etc.  The latter probably
+		//                works for all cases, so transactions are probably not necessary.
+		//                Actually, on further thought, the race condition exists anyway - neither of the above tow really solve it.
+		//                Ho hum, have to think harder.
+		s.eventRecorder.Event(service, api.EventTypeNormal, "CreatingDNSRecords", "Creating DNS Records")
+		err := s.ensureDNSRecords(clusterName, service)
+		if err != nil {
+			return fmt.Errorf("Failed to create DNS records for service %s: %v", key, err), retryable
```

Comment:
This function returns two variables not one?  I don't think that this will compile.